crates/valence_command/src/arg_parser.rs

```diff
@@ -0,0 +1,970 @@
+    type Result = String;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(string)
+    }
+
+    fn display() -> Parser {
+        Parser::BlockPredicate
+    }
+}
+
+// TODO: item stack proper
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct ItemStack;
+
+impl CommandArg for ItemStack {
+    type Result = String;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(string)
+    }
+
+    fn display() -> Parser {
+        Parser::ItemStack
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum ChatColor {
+    Black,
+    DarkBlue,
+    DarkGreen,
+    DarkAqua,
+    DarkRed,
+    DarkPurple,
+    Gold,
+    Gray,
+    DarkGray,
+    Blue,
+    Green,
+    Aqua,
+    Red,
+    LightPurple,
+    Yellow,
+    #[default]
+    White,
+    Reset,
+}
+
+impl CommandArg for ChatColor {
+    type Result = ChatColor;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(match string.to_lowercase().as_str() {
+            "black" => ChatColor::Black,
+            "dark_blue" => ChatColor::DarkBlue,
+            "dark_green" => ChatColor::DarkGreen,
+            "dark_aqua" => ChatColor::DarkAqua,
+            "dark_red" => ChatColor::DarkRed,
+            "dark_purple" => ChatColor::DarkPurple,
+            "gold" => ChatColor::Gold,
+            "gray" => ChatColor::Gray,
+            "dark_gray" => ChatColor::DarkGray,
+            "blue" => ChatColor::Blue,
+            "green" => ChatColor::Green,
+            "aqua" => ChatColor::Aqua,
+            "red" => ChatColor::Red,
+            "light_purple" => ChatColor::LightPurple,
+            "yellow" => ChatColor::Yellow,
+            "white" => ChatColor::White,
+            "reset" => ChatColor::Reset,
+            _ => {
+                return Err(CommandArgParseError::InvalidArgument(
+                    "chat_color".to_string(),
+                    string,
+                ))
+            }
+        })
+    }
+
+    fn display() -> Parser {
+        Parser::Color
+    }
+}
+
+// TODO: json chat component proper
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct JsonChatComponent;
+
+impl CommandArg for JsonChatComponent {
+    type Result = String;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(string)
+    }
+
+    fn display() -> Parser {
+        Parser::Component
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct Message;
+
+impl CommandArg for Message {
+    type Result = String;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(string)
+    }
+
+    fn display() -> Parser {
+        Parser::Message
+    }
+}
+
+// TODO: nbt proper
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct Nbt;
+
+impl CommandArg for Nbt {
```

Comment:
Could use the `valence_server::nbt::Compound` type. We have a (somewhat incomplete) SNBT encoders/decoders there.